modules/c++/xml.lite/source/Speed.cpp

```diff
@@ -0,0 +1,281 @@
+  std::map<std::string, size_t> local_registry;
+  // either both or niether parent and element must be specified
+  if (parent.empty() ^ (element == NULL))
+  {
+    throw except::Exception(
+        "xml::lite::Speed::buildTOC() called with only one of"
+        " parent path or current element."
+        " Only neither or both are allowed.");
+  }
+
+  if (element == NULL)
+  {
+    element = mRootNode;
+    // If we are starting at the top, clear out the old TOC before rebuilding.
+    mTOC.clear();
+  }
+
+  std::vector<xml::lite::Element *> childs = element->getChildren();
+  size_t sz = childs.size();
+  for (size_t ii = 0; ii < sz; ii++)
+  {
+    std::string lname = childs[ii]->getLocalName();
+    if (local_registry.count(lname) == 0)
+    {
+      local_registry[lname] = 1;
+    }
+    else
+    {
+      ++local_registry[lname];
+    }
+    std::string toc_key = parent+">"
+                          +lname+"&"
+                          +str::toString<size_t>(local_registry[lname]);
+    mTOC[toc_key] = childs[ii];
+    buildTOC(toc_key, childs[ii]);
+  }
+
+}
+//=================================================================
+std::string xml::lite::Speed::getData(const std::string &elname)
+{
+  return mTOC[elname]->getCharacterData();
+}
+//=================================================================
+void xml::lite::Speed::initRoot(const std::string &rootName)
+{
+  setRootElement(createElement(rootName,"",""));
+}
+//=================================================================
+xml::lite::Speed::KeyParts xml::lite::Speed::extractParts(const std::string &longname)
+{
+  xml::lite::Speed::KeyParts parts;
+  // make this name a proper name by ensuring zero-ordinal names have '&1' 
+  std::vector<std::string> chunks = str::split(longname, ">");
+
+  size_t ord_del_idx;
+  for(size_t ii = 0; ii<chunks.size(); ++ii)
+  { 
+    ord_del_idx = chunks[ii].find_last_of("&");
+    if (ord_del_idx == std::string::npos) chunks[ii] += "&1";
+  }
+  
+  parts.proper_name = str::join(chunks, ">");
+
+  size_t delim_idx = parts.proper_name.find_last_of('>');
+  // grab string up to but not including the last '>' delimiter
+  if (delim_idx != std::string::npos)
+  {
+    parts.parent = parts.proper_name.substr(0, delim_idx);
+  }
+  else
+  {
+    parts.parent = "";
+  }
+  // grab the string of everything past the last '>' delimiter
+  // the +1 works here even if delim_idx is std::string::npos, since that
+  // value is essentially -1.
+  parts.child = parts.proper_name.substr(delim_idx+1, std::string::npos);
+
+  size_t ord_delim_idx = parts.child.find_last_of('&');
+  parts.child_name = parts.child.substr(0, ord_delim_idx);
+  if (ord_delim_idx != std::string::npos)
+  {
+    parts.child_ord = atoi(parts.child.substr(ord_delim_idx+1).c_str());
+  }
+  else
+  {
+    // this branch should never happen- above code ensures presence of
+    // ordinal delimiter.
+    std::cout<<"NOT PROPER NAME! :"<<parts.child<<std::endl;
```

Comment:
If this should never happen, shouldn't this throw?  It shouldn't print... needs to either throw or we could pass in a logger if this is somehow just a warning